File: nautical/io/buoy.py
from __future__ import annotations

from re import sub

from bs4 import BeautifulSoup

from nautical.cache.file import load_buoy, save_buoy
from nautical.cache.memory import _buoy_cache
from nautical.io.web import get_noaa_forecast_url, get_url_source
from nautical.log import get_logger
from nautical.noaa.buoy.buoy import Buoy
from nautical.noaa.buoy.buoy_data import BuoyData

log = get_logger(__name__)


def create_buoy(buoy: str, use_cache: bool = True) -> Buoy | None:
    """Provide a full workup for a specific buoy. If the buoy is None or it cannot
    be found then the data returned will be considered invalid as None

    :param buoy: id of the buoy to do a workup on
    :param use_cache: when True, check memory and disk caches before fetching
    :return: BuoyWorkup if successful else None
    """
    if not buoy:
        return None

    if use_cache:
        cached = _buoy_cache.get(buoy)
        if cached is not None and isinstance(cached, Buoy) and cached.valid:
            return cached

        disk_cached = load_buoy(buoy)
        if disk_cached is not None and disk_cached.valid:
            _buoy_cache.set(buoy, disk_cached)
            return disk_cached

    buoy_data = Buoy(buoy)
    _fill_buoy_from_network(buoy_data)

    if not buoy_data.valid:
        return None

    if use_cache:
        _buoy_cache.set(buoy, buoy_data)
        try:
            save_buoy(buoy_data)
        except OSError:
            log.warning("Failed to write buoy %s to disk cache", buoy)

    return buoy_data


def fill_buoy(buoy: Buoy, use_cache: bool = True) -> None:
    """Pass in a Buoy object that needs to be filled in with the current data.
    The buoy object will have the validity set if the results were successful

    :param buoy: nautical.noaa.buoy.Buoy object
    :param use_cache: when True, check memory and disk caches before fetching
    """
    if use_cache:
        cached = _buoy_cache.get(buoy.station)
        if cached is not None and isinstance(cached, Buoy) and cached.valid:
            buoy.valid = cached.valid
            buoy.present = cached.present
            return

        disk_cached = load_buoy(buoy.station)
        if disk_cached is not None and disk_cached.valid:
            buoy.valid = disk_cached.valid
            buoy.present = disk_cached.present
            _buoy_cache.set(buoy.station, buoy)
            return

    _fill_buoy_from_network(buoy)

    if use_cache and buoy.valid:
        _buoy_cache.set(buoy.station, buoy)
        try:
            save_buoy(buoy)
        except OSError:
            log.warning("Failed to write buoy %s to disk cache", buoy.station)


def _fill_buoy_from_network(buoy: Buoy) -> None:
    """Fetch buoy data from NOAA and populate the buoy object."""
    url = get_noaa_forecast_url(buoy.station)
    soup = get_url_source(url)

    current_buoy_data = BuoyData()
    buoy_valid = get_current_data(
        soup, current_buoy_data, [f"Conditions at {buoy.station}", "Detailed Wave Summary"]
    )
    buoy.valid = buoy_valid
    buoy.present = current_buoy_data


def get_current_data(soup: BeautifulSoup, buoy: BuoyData, search: str | list[str]) -> bool:
    """Search the beautiful soup object for a TABLE containing the search string. The
    function will grab the data from the table and create a NOAAData object and return the data

    :param soup: beautiful soup object generated from the get_url_source()
    :param buoy: BuoyData object that should be filled with data as this function parses the data.
    :param search: text to search for in the soup object.
    :return: True when data has been found and set
    """
    # keep track of the number of variables that were set, indicates validity
    buoy_variables_set = 0

    if not isinstance(search, list):
        search = [search]

    # Find all tables with a caption that has the text we are searching for
    tables = []
    for caption in soup.find_all("caption"):
        for search_text in search:
            if search_text in caption.get_text():
                tables.append(caption.find_parent("table"))

    for table in tables:
        for i, row in enumerate(table.find_all("tr")):
            key_data = None
            key = None
            value = None

            # the first table is another table and it is no use to use -- skipping
            if i >= 1:
                cells = row.find_all("td")
                if cells:
                    try:
                        key_data = cells[0].text
                        key = key_data[key_data.find("(") + 1 : key_data.find(")")]
                        value = cells[1].next.split()[0]
                        buoy.set(key.lower(), value)
                        buoy_variables_set += 1
                    except (IndexError, TypeError, AttributeError) as error:
                        log.error(
                            "%s - key_field: %s, key: %s, value: %s",
                            error,
                            key_data,
                            key,
                            value,
                        )

    # no variables set indicates errors or invalid buoy
    return buoy_variables_set > 0


# Alias for getting current data, It has the same result
get_buoy_data = get_current_data

File: pkg/noaa/buoy/buoy.go
package buoy

import (
	"encoding/json"
	"fmt"
	"hash/fnv"
	"regexp"
	"strconv"
	"strings"

	"github.com/anaskhan96/soup"

	"github.com/barbacbd/nautical/pkg/io"
	"github.com/barbacbd/nautical/pkg/location"
	"github.com/barbacbd/nautical/pkg/util"
)

var (
	// NauticalRegex is a regular expression to find values between the parentheses
	NauticalRegex = regexp.MustCompile(`\((.*?)\)`)

	// aliases is a set of all alias names for variables of a buoy
	aliases = util.GetAliases(BuoyData{})
)


// Buoy represents a NOAA buoy.
type Buoy struct {
	// Station is the station ID or ID of the buoy
	Station string `json:"station"`

	// Description is the string description for the Buoy
	// +optional
	Description string `json:"description,omitempty"`

	// Location is the geographical location of the buoy
	// +optional
	Location location.Point `json:"location,omitempty"`

	// Present is the current BuoyData associated with this Buoy
	// +optional
	Present *BuoyData `json:"data,omitempty"`

	// Past is a list of BuoyData structs that are from the Past
	// Deprecated: the past variable is still provided for user assistance, but
	// the value is considered deprecated and abandoned
	// +optional
	Past []*BuoyData `json:"past,omitempty"`

	// Determines whether the data stored in this stuct is considered valid
	// +optional
	Valid bool `json:"valid,omitempty"`
}

// Placemark is the xml structure for a Buoy. It is a very different
// format than the normal Buoy
type Placemark struct {
	// Name is the name tag in the Placemark that should match the
	// name of a Buoy
	Name string `xml:"name"`
	// Description is the description tag of the Placemark that should
	// match the description of the Buoy
	Description string `xml:"description"`
	// Placement is the location of the buoy
	Placement location.Point `xml:"LookAt"`
}

// Hash returns the hashed string as an integer from Buoy
func (b *Buoy) Hash() uint64 {
	hash := fnv.New64a()

	var hashString string

	if b.Description != "" {
		hashString = fmt.Sprintf("%s %s", b.Station, b.Description)
	} else {
		hashString = b.Station
	}

	hash.Write([]byte(hashString))

	return hash.Sum64()
}

// SetData sets the current data and moves the current data to the past
func (b *Buoy) SetData(data *BuoyData) error {
	if b.Present != nil {

		dataEpoch, err := data.EpochTime()
		if err != nil {
			return nil
		}

		presentEpoch, err := b.Present.EpochTime()
		if err != nil {
			return nil
		}

		if dataEpoch <= presentEpoch {
			return fmt.Errorf("failed to set data, epoch time is older than present data")
		}

		b.Past = append(b.Past, b.Present)
	}

	b.Present = data
	return nil
}

// CreateBuoy will provide a full workup for a specific buoy. When the
// buoy with matching station is not found, an error is returned.
func CreateBuoy(stationID string) (*Buoy, error) {
	station := &Buoy{Station: stationID}

	err := station.FillBuoy()
	if err != nil {
		return nil, err
	}

	if !station.Valid {
		return nil, fmt.Errorf("buoy %s is not valid", stationID)
	}

	return station, nil
}

// FillBuoy will fill a Buoy struct with the data parsed from the web
func (b *Buoy) FillBuoy() error {
	url := io.GetNOAAForecastURL(b.Station)
	root, err := io.GetURLSource(url)
	if err != nil {
		return err
	}

	search := []string{fmt.Sprintf("Conditions at %s", b.Station), "Detailed Wave Summary"}
	if err := b.GetCurrentData(root, search); err != nil {
		return err
	}
	return nil
}

// GetCurrentData parses the current data for a buoy from the NOAA website
func (b *Buoy) GetCurrentData(root *soup.Root, search []string) error {
	buoyVariablesSet := false

	tables := map[string]soup.Root{}

	// Find all tables that have a caption that matches any of
	// the search criteria
	suspectTables := root.FindAll("table")
	for _, suspect := range suspectTables {
		captions := suspect.FindAll("caption")
		captionText := []string{}
		for _, caption := range captions {
			captionText = append(captionText, caption.Text())
		}
		for _, searchText := range search {
			for _, ct := range captionText {
				if strings.Contains(ct, searchText) {
					if _, ok := tables[ct]; !ok {
						tables[ct] = suspect
					}

				}
			}
		}
	}

	for _, table := range tables {
		allTR := table.FindAll("tr")
		for i, row := range allTR {
			if i >= 1 {
				cells := row.FindAll("td")
				if len(cells) >= 2 {
					// for idx, cell := range cells {
					submatchall := NauticalRegex.FindAllString(cells[0].Text(), -1)
					if len(submatchall) > 0 {
						// Trim the () off of the value so that we can use the variable name
						alias := strings.Trim(strings.Trim(strings.ToLower(submatchall[0]), "("), ")")

						// Make sure that this is data that we are expecting
						if aliases.Has(alias) {
							splitCell := RemoveEmpty(strings.Split(cells[1].Text(), " "))
							val, err := strconv.ParseFloat(splitCell[0], 64)
							if err != nil {
								json.Unmarshal([]byte(fmt.Sprintf("{\"%s\": \"%s\"}", alias, splitCell[0])), b.Present)
							} else {
								json.Unmarshal([]byte(fmt.Sprintf("{\"%s\": %f}", alias, val)), b.Present)
							}
							buoyVariablesSet = true
						}
					}
				}
			}
		}
	}

	b.Valid = buoyVariablesSet
	if !buoyVariablesSet {
		return fmt.Errorf("no buoy variables set")
	}
	return nil
}
